### PR TITLE
[scan] Autograd

### PR DESCRIPTION
**Pull request by bohnstingl**
This is the new PR to realize autograd for scan. This is a derivative of https://github.com/bohnstingl/pytorch/pull/2

@ydwu4 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -6745,7 +6745,7 @@ def extract_scan_args(
         _, fx_init, fx_xs, _, _, fx_additional_inputs = extract_scan_args(
             *V.graph.current_node.args
         )
-        fx_all_inputs = fx_init + fx_xs + fx_additional_inputs
+        fx_all_inputs = fx_init + fx_xs + tuple(fx_additional_inputs) if type(fx_init) == tuple else fx_additional_inputs
```
Comment (Skylion007) [inline torch/_inductor/ir.py:6748]:
```suggestion
        fx_all_inputs = fx_init + fx_xs + tuple(fx_additional_inputs) if type(fx_init) is tuple else fx_additional_inputs
```

**Review by ydwu4 — COMMENTED** [review]
Overall we're on the right track! Add a few comments. The additonal_inputs would need some special handling and some more testing.

File: torch/_higher_order_ops/scan.py [hunk 2]
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+def scan_autograd(combine_fn, init, input, dim, reverse, additional_inputs):
+    # A shortcut for the case where all inputs don't require gradient,
+    # we skip tracing the forward and backward graph.
+    if pytree.tree_all_only(
+        torch.Tensor,
+        lambda t: not t.requires_grad,  # type: ignore[union-attr]
+        (init, input),
+    ):
+        with torch._C._AutoDispatchBelowAutograd():
+            return scan_op(combine_fn, init, input, dim, reverse, additional_inputs)
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:541]: Currently it does, I can see this in test. But if this the DispatchKey is not invoked at all, this would also be an option

File: torch/_higher_order_ops/scan.py [hunk 3]
```diff
@@ -42,9 +58,62 @@ def wrap_combine_fn_flat(
     assert num_init_leaves == len(carry_flat)
     return [*carry_flat, *combined_flat]
 
+def create_fw_bw_graph_combinefn(combine_fn, init, input, dim):
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_higher_order_ops/scan.py [hunk 4]
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+        num_leaves_init = ctx._num_leaves_init
+        num_leaves_xs = ctx._num_leaves_xs
+        reverse = ctx._reverse
+
+        # Retrieve the forward inputs and the forward outputs
+        operands_outs = ctx.saved_tensors
+        init = operands_outs[:num_leaves_init]
+        xs = operands_outs[num_leaves_init : num_leaves_init + num_leaves_xs]
+        carries = operands_outs[
+            num_leaves_init + num_leaves_xs : 2 * num_leaves_init + num_leaves_xs
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:554]: Done

File: torch/_higher_order_ops/scan.py [hunk 5]
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+            if reverse:
+                carries_augmented = [
+                    torch.cat(
+                        [torch.unsqueeze(i, dim), torch.flip(c[1:], [dim])], dim=dim
+                    )
+                    for i, c in zip(init, carries)
+                ]
+                xs = [torch.flip(x, [dim]) for x in xs]
+            else:
+                carries_augmented = [
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_higher_order_ops/scan.py [hunk 6]
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+                ]
+
+            xs_bwd = (*g_ys, *carries_augmented, *xs)
+            g_init, g_xs = _extract_carry_and_out(
+                scan_op(joint_graph, g_c_T, xs_bwd, dim, True, additional_inputs),
+                num_leaves_init,
+            )
+
+            if reverse:
+                g_xs = [torch.flip(g, [dim]) for g in g_xs]
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_inductor/ir.py (hunk 1 above)
Comment (bohnstingl) [reply torch/_inductor/ir.py]: This has been resolved in a different way

File: torch/_higher_order_ops/scan.py [hunk 7]
```diff
@@ -360,9 +410,153 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+    ):
+        with torch._C._AutoDispatchBelowAutograd():
+            return scan_op(combine_fn, init, input, dim, reverse, additional_inputs)
+
+    num_leaves_init = len(init)
+
+    (
+        fw_graph,
+        joint_graph,
+    ) = create_fw_bw_graph_combinefn(combine_fn, init, input, dim)
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: I implemented a version of this and also modified the RNN testcase to demonstrate that

File: torch/_higher_order_ops/scan.py [hunk 8]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+
+
+@scan_op.py_impl(DispatchKey.Autograd)
+def scan_autograd(combine_fn, init, xs, dim, reverse, additional_inputs):
+    # A shortcut for the case where all inputs don't require gradient,
+    # we skip tracing the forward and backward graph.
+    if pytree.tree_all_only(
+        torch.Tensor,
+        lambda t: not t.requires_grad,  # type: ignore[union-attr]
+        (init, xs),
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:602]: We also need to check additional_inputs?

File: torch/_higher_order_ops/scan.py [hunk 9]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+    ):
+        with torch._C._AutoDispatchBelowAutograd():
+            return scan_op(combine_fn, init, xs, dim, reverse, additional_inputs)
+
+    num_leaves_init = len(init)
+    num_leaves_additional_inputs = len(additional_inputs)
+
+    if not torch.is_grad_enabled():
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:610]: This worths a comment on when this could happen: e.g. in the double-backward case when we trigger the autogradKey of backward. Can merge with the previous TODO.

probably use a different variable name like new_dim and add a comment on why we force it to be 0.

File: torch/_higher_order_ops/scan.py [hunk 10]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+        dim,
+        reverse,
+        num_leaves_init,
+        num_leaves_additional_inputs,
+        *ops,
+    ):
+        ops_list = list(ops)
+        init = ops_list[:num_leaves_init]
+        xs = ops_list[num_leaves_init : len(ops_list) - num_leaves_additional_inputs]
+        additional_inputs = ops_list[len(ops_list) - num_leaves_additional_inputs :]
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:474]: Maybe a locally scoped helper function:
`def split_operands(ops, num_leaves_init, num_leaves_additional_inputs)`

nit: ops usually refers to operators, so probably change it flat_args or something

File: torch/_higher_order_ops/scan.py [hunk 11]
```diff
@@ -4,32 +4,93 @@
...
+def shift_source_dim_to_target_dim(t, from_dim: int, to_dim: int):
+    assert to_dim >= 0 and to_dim < t.ndim
+    assert from_dim >= 0 and from_dim < t.ndim
+    dims = list(range(0, t.ndim))
+    dims.pop(from_dim)
+    dims.insert(to_dim, from_dim)
+    return t.permute(*dims)
+
+
+def prepare_xs_carries_for_bwd(xs, init, carries, dim, reverse):
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:68]: We don't expect two prepare_* functions to be used by other file, right? So maybe move these two prepare function inside of backward to make it locally scoped? They are not exposed to users in this way.

File: torch/_higher_order_ops/scan.py [hunk 12]
```diff
@@ -43,8 +104,64 @@ def wrap_combine_fn_flat(
     return [*carry_flat, *combined_flat]
 
 
-def _extract_carry_and_out(flat_out: List[Any], num_carry: int):
-    return flat_out[:num_carry], flat_out[num_carry:]
+def create_fw_bw_graph_combinefn(combine_fn, init, xs, dim, additional_inputs):
+    # See Note [HOP create fw_bw graph] in create_fw_bw_graph in utils.py
+
+    # Helper wrapper for the autograd forward.
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:110]: Can add a comment 
```
# TODO: do less re-computation with min-cut partitioner. 
```

File: torch/_higher_order_ops/scan.py [hunk 13]
```diff
@@ -43,8 +104,64 @@ def wrap_combine_fn_flat(
...
+                    "Expect carry produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(carry) for carry in fw_carry]}."
+                )
+            if any(not isinstance(out, torch.Tensor) for out in fw_outputs):
+                raise RuntimeError(
+                    "Expect outputs produced by combine_fn to only contains tensors. "
+                    f"Got types {[type(out) for out in fw_outputs]}."
+                )
+
+            # TODO: There is a major issue that the create_fw_bw in the higher_order_op is invoked twice:
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:148]: We can probably move to where we do the torch.is_grad_enabled() check (need to change the wording a bit).

File: torch/_higher_order_ops/utils.py [hunk 14]
```diff
@@ -264,10 +264,12 @@ def maybe_clone(t):
 def prepare_fw_with_masks(fn):
     def fw_with_masks(*args):
         fw_out = fn(*args)
-        return fw_out, [
+        fw_out_flat, tree_spec = pytree.tree_flatten(fw_out)
```
Comment (ydwu4) [inline torch/_higher_order_ops/utils.py:267]: Do we need this change?

File: torch/_higher_order_ops/scan.py [hunk 15]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+            return (*carry, *outs)
+
+    @staticmethod
+    def backward(ctx, *flat_grads):
+        r"""
+        This function computes the gradients of the scan operation.
+        It does so by constructing using an additional scan operator with the gradients
+
+        Args:
+            flat_grads (torch.Tensor): The tensor of upstream gradients, or a nested pytree of tensors.
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:504]: I believe they're always flattened  with no pytree involved.

File: torch/_higher_order_ops/scan.py [hunk 16]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+
+        Example::
+
+            The ``fw_graph`` f(.,.), used in the forward function, is the operator used during the scan. For example
+            def f(x: torch.Tensor, y: torch.Tensor):
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:509]: this example is a bit boring: maybe change to next_carry = y = x * y?

File: torch/_higher_order_ops/scan.py [hunk 17]
```diff
@@ -110,9 +227,6 @@ def add(x: torch.Tensor, y: torch.Tensor):
     if not isinstance(reverse, bool):
         raise RuntimeError("Reverse must be a bool, but got " + str(type(reverse)))
 
-    # TODO: Support closures/nn_modules in order to be able represent RNNs with scan
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:113]: Don't need to happen in this PR: it seems to me that can we handle the dim kwarg in scan (i.e. the user-facing wrapper) by shift it backward to dim 0 so that we could remove all the dim arguments in scan_op. This would simplify everything related to dim. WDYT? Maybe add a TODO?

**Review by ydwu4 — APPROVED** [review]

File: test/functorch/test_control_flow.py [hunk 18]
```diff
@@ -2459,8 +2527,16 @@ def add_scalar_carry(x: torch.Tensor, y: torch.Tensor):
         self.assertEqual(result_init, result_exp)
         self.assertEqual(result_init[0], torch.tensor([3.0], device=device))
 
+        if autograd:
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:2530]: all the autograd tests looks similar, can we extract them out as a funciton: check_autograd?

File: torch/_higher_order_ops/scan.py [hunk 19]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+                ),
+                num_leaves_init,
+            )
+
+            # g_xs may contain the gradients for the additional inputs as well which need to be separated
+            new_g_additional_inputs = g_xs[num_leaves_xs:] + old_g_additional_inputs_T
+            g_xs = g_xs[:num_leaves_xs]
+            g_xs = prepare_final_gradients_xs(g_xs, dim, reverse)
+
+        return *[None] * 6, *g_init, *g_xs, *new_g_additional_inputs
```
Comment (ydwu4) [inline torch/_higher_order_ops/scan.py:592]: Do we have tests for testing gradient of additional_inputs? I feel we need to modify the backward graph as well:
```
def new_joint_graph(carried_grad_additional_inputs, ...):
   ...grad_additioal_inptus... = joint_graph(...)
   carry (..., grad_addtioanl_inputs + carried_grad_additional_inputs),
```
The init should be a torch.zeros() note that: the joint_graph doesn't take carried_grad_additional_inputs as carry. it only takes additioanl_inputs as part of xs. It's the wrapper new_joint_graph that takes a new carry.

File: test/functorch/test_control_flow.py (hunk 18 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py:2531]: I factored out most of them

File: torch/_higher_order_ops/scan.py (hunk 17 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:113]: I am not so sure, it may be that the `combine_fn` depends internally on the shape that the user provides, right?
One concrete example:
```
W = torch.randn(3, 5)
xs = torch.randn(5, 7, 3)
init = torch.randn(3, 3)
def f(x, y):
  return W @ x + y, W @ x + y
carry, out = scan(f, init, xs, 1)
```
In this case we cannot always move the dim to 0, as the `matmul` would break, right?

File: torch/_higher_order_ops/scan.py (hunk 16 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:498]: I rewrote the example for the multiply function

File: torch/_higher_order_ops/scan.py (hunk 15 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Corrected

File: torch/_higher_order_ops/utils.py (hunk 14 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py]: Good catch, no we don't. This was an artifact from resolving earlier issues

File: torch/_higher_order_ops/scan.py (hunk 13 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_higher_order_ops/scan.py (hunk 12 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:85]: Done

File: torch/_higher_order_ops/scan.py (hunk 11 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: I moved the functions inside the backward

File: torch/_higher_order_ops/scan.py (hunk 10 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done, I simplified and unified the code

File: torch/_higher_order_ops/scan.py (hunk 9 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_higher_order_ops/scan.py (hunk 8 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: Done

File: torch/_higher_order_ops/scan.py (hunk 19 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:647]: I've included a wrapper code that takes care of the gradients for the additional inputs

File: torch/_higher_order_ops/scan.py [hunk 20]
```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
...
+            num_leaves_init + num_leaves_xs : 2 * num_leaves_init + num_leaves_xs
+        ]
+        additional_inputs = list(operands_outs[2 * num_leaves_init + num_leaves_xs :])
+
+        with torch._C._AutoDispatchBelowAutograd():
+            # The flat gradients are a list of g_c_T, g_ys and optionally the gradients for the additional_inputs
+            flat_grads_list = list(flat_grads)
+            g_c_T = flat_grads_list[:num_leaves_init]
+            g_ys = flat_grads_list[num_leaves_init : num_leaves_init + num_leaves_ys]
+            old_g_additional_inputs_T = flat_grads_list[
```
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: I've added an additional test to ensure that the gradients for the additional inputs are correct

File: torch/_higher_order_ops/scan.py (hunk 17 above)
Comment (ydwu4) [reply torch/_higher_order_ops/scan.py:113]: like we could do scan(f, init, xs_shifted, 0), where xs_shifted.shape = (7, 5, 3)

Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py:113]: True, I'll add a TODO for that

File: torch/_higher_order_ops/scan.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bohnstingl) [reply torch/_higher_order_ops/scan.py]: I added a TODO for this

**Comment by bohnstingl** [comment]
This PR will be combined with https://github.com/bohnstingl/pytorch/pull/4 and pursued further in https://github.com/pytorch/pytorch/pull/146285
